scripts/deno/createMainChangelog.ts

```diff
@@ -0,0 +1,207 @@
+import { readFileSync, writeFileSync } from "node:fs";
+import { List, ListItem, Root, Text } from "npm:@types/mdast";
+import { Node } from "npm:@types/unist";
+import { heading, root, text } from "npm:mdast-builder";
+import remarkParse from "npm:remark-parse";
+import remarkStringify from "npm:remark-stringify";
+import { unified } from "npm:unified";
+import { EXIT, SKIP, visit } from "npm:unist-util-visit";
+import { visitParents } from "npm:unist-util-visit-parents";
+import { getChangelogs } from "./utils.ts";
+
+/**
+ * Small diagram of the process:
+
+           original 
+        markdown files
+               |
+               V
+          custom JSON 
+        representation
+         of changelog
+               |
+               V
+        final markdown
+ */
+
+type PackageName = string;
+type Version = string;
+type RawASTNode = Node;
+type SemverChangeHeading = Record<string, RawASTNode[]>;
+type VersionEntry = Record<PackageName, SemverChangeHeading>;
+type Changelog = Record<Version, VersionEntry>;
+
+const upsertEntry = (
+  changelog: Changelog,
+  {
+    lastSeenPackage,
+    lastSeenVersion,
+    lastSeenSemverHeading,
+  }: {
+    lastSeenPackage: string;
+    lastSeenVersion: string;
+    lastSeenSemverHeading: string;
+  },
+  value: RawASTNode
+) => {
+  if (!changelog[lastSeenVersion]) {
+    changelog[lastSeenVersion] = {};
+  }
+  if (!changelog[lastSeenVersion][lastSeenPackage]) {
+    changelog[lastSeenVersion][lastSeenPackage] = {};
+  }
+  const raw_ast_nodes =
+    changelog[lastSeenVersion][lastSeenPackage][lastSeenSemverHeading];
+  if (!raw_ast_nodes) {
+    changelog[lastSeenVersion][lastSeenPackage][lastSeenSemverHeading] = [
+      value,
+    ];
+  } else {
+    raw_ast_nodes.push(value);
+  }
+
+  const updatedVersion = changelog[lastSeenVersion] || [];
+  changelog[lastSeenVersion] = updatedVersion;
+};
+
+const parseMarkdownFiles = async (filePaths: string[]): Promise<Changelog> => {
+  const changelog: Changelog = {};
+
+  for (const filePath of filePaths) {
+    const fileContent = readFileSync(filePath, { encoding: "utf-8" });
+    const fileAST = await unified().use(remarkParse).parse(fileContent);
+
+    // TODO is there a _better_ way? This seems cluttered and hard to reason about
+    let lastSeenPackage = "";
```

Comment:
Eneste jeg ser for meg i farten er å potensielt gjøre en preprossessering av AST-en før man begynner å hente ut dataen. Kan da gjøre noen antagelser som at depth-1 er pakkanavn, depth-2 er versjon etc

Kanskje noe lignende som dette? 
```ts

groupAstByVersion(fileAST.children);

type VersionCollection = { version: string; nodes: Node[] }[];

const groupAstByVersion = (
  astNode: RawASTNode[]
): { name: string; versions: VersionCollection } => {
  // antar at første node alltid er h1
  const name = astNode[0].value;
  const versions: VersionCollection = [];
  let currentVersion = { version: "", nodes: [] as Node[] };
  astNode.slice(1).forEach((node) => {
    if (node.type === "heading" && node.depth === 2) {
      if (currentVersion.version.length > 0) {
        versions.push(structuredClone(currentVersion));
        currentVersion = { version: "", nodes: [] };
      }
      currentVersion.version = node.value;
    } else {
      currentVersion.nodes.push(node);
    }
  });

  return { name, versions };
};
```